### PR TITLE
[string] Chunk reads

### PR DESCRIPTION
**Pull request by faho**
## Description

Profiling with callgrind revealed that about 60% of the time in a `something | string match` call
was actually spent in `string_get_arg_stdin()`,
because it was calling `read` one byte at a time.

This makes it read in chunks similar to builtin read.

This increases performance for `getent hosts | string match -v '0.0.0.0*'` from about 300ms to about 30ms (i.e. 90%).
At that point it's _actually_ quicker than `grep`.

To improve performance even more, we'd have to cut down on str2wcstring.

Fixes issue #4604.

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

**Comment by faho** [comment]
Urgh, I need to attribute https://stackoverflow.com/questions/1583353/how-to-read-exactly-one-line/1584620#1584620, which this is based on.

File: src/builtin_string.cpp [hunk 1]
```diff
@@ -58,31 +59,33 @@ static bool string_args_from_stdin(const io_streams_t &streams) {
...
-            if (arg.empty()) {
-                return NULL;
-            }
-            break;
+    // We might read more than a line - store the rest in a static buffer.
+    static std::string buffer;
+
+    // Read in chunks from fd until buffer has a line.
+    std::string::iterator pos;
+    while ((pos = std::find (buffer.begin(), buffer.end(), '\n')) == buffer.end ()) {
```
Comment (ridiculousfish) [inline src/builtin_string.cpp]: This is a bit more idiomatic as:
```
    size_t pos;
    while ((pos = buffer.find('\n')) == std::string::npos) {...
```

File: src/builtin_string.cpp [hunk 2]
```diff
@@ -58,31 +59,33 @@ static bool string_args_from_stdin(const io_streams_t &streams) {
...
         }
-
-        arg += ch;
+        buffer.append(buf, n);
     }
 
-    *storage = str2wcstring(arg);
-    return storage->c_str();
+  // Split the buffer around '\n' found and return first part.
+  *storage = str2wcstring(buffer.c_str(), std::distance(buffer.begin(), pos));
```
Comment (ridiculousfish) [inline src/builtin_string.cpp]: With the above this could be simply `*storage = str2wcstring(buffer, pos);`

File: src/builtin_string.cpp [hunk 3]
```diff
@@ -58,31 +59,33 @@ static bool string_args_from_stdin(const io_streams_t &streams) {
...
+  buffer = std::string(pos + 1, buffer.end());
```
Comment (ridiculousfish) [inline src/builtin_string.cpp]: and this could be `buffer.erase(0, pos);`

**Comment by ridiculousfish** [comment]
I think with my suggestions we can skip the attributions

File: src/builtin_string.cpp (hunk 1 above)
Comment (faho) [reply src/builtin_string.cpp]: THAT WAS IT! I've been thinking that there was some nicer way to to `find`, but I couldn't for the life of me come up with it.

**Comment by faho** [comment]
Merged with a bit of cleanup as 2de38ef7bfd8569a6876d8b8ffbe748eec1b3c43, with some additional work in f9d883dd745064650669ea48d6520c48b54537d6.

**Pull request closed without merging**